src/modules/ctc/inc/ctc.h

```diff
@@ -7,28 +7,162 @@
 #include "wayside_controller.h"
 #include "wayside_controller_gateway.h"
+#include "graph.h"
 
 namespace ctc
 {
 
-typedef struct
+typedef uint16_t                         column;
+typedef struct DestinationAndArrivalTime DestinationAndArrivalTime;
+typedef struct Train                     Train;
+typedef struct Station                   Station;
+
+typedef enum
+{
+    SectionDirection_Forward,
+    SectionDirection_Reverse,
+} SectionDirection;
+
+typedef enum
+{
+    AUTOMATIC_MODE,
+    MANUAL_MODE,
+    MAINTENANCE_MODE
+} CtcOperationMode;
+
+struct Station
+{
+    Station(void) : station_name(""), block_id(0)
+    {
+    };
+    Station(std::string station_name, types::BlockId block_id, std::chrono::duration<double> t) : station_name(station_name), block_id(block_id), total_time_to_station_with_dwell(t)
+    {
+    };
+    std::string station_name;
+    types::BlockId block_id;
+    std::chrono::duration<double> total_time_to_station_with_dwell;
+};
+
+struct DestinationAndArrivalTime
 {
-    types::BlockId block;
+    DestinationAndArrivalTime(void);
+    DestinationAndArrivalTime(const Station station, types::Tick arrival_time) : station(station), arrival_time(arrival_time)
+    {
+    };
+    ctc::Station station;
     types::Tick arrival_time;
-} DestinationAndArrivalTime;
+};
+
+class ScheduleBuilder
+{
+    public:
+        ScheduleBuilder(void);
+        ScheduleBuilder(std::filesystem::path &file_path);
+};
+
+class GraphBuilder
+{
+    public:
+        GraphBuilder(void);
+        GraphBuilder(std::vector<types::Block> &blocks);
+        void BuildGreenLine(void);
+        std::vector<types::Block> BuildSection(types::SectionId section_id);
+        void AddSection(std::vector<types::Block> section, types::BlockDirection direction);
+
+        std::size_t GetBlockList(void) const;
+        types::BlockId GetSectionBegin(void) const;
+        types::BlockId GetSectionEnd(void) const;
+        Graph<types::BlockId, types::Meters> GetGraph(void);
+
+    private:
+        std::vector<types::Block> block_list_;
+        Graph<types::BlockId, types::Meters> graph_;
+};
+
+struct Train
+{
+    Train(void) : train_id(GetNextId())
+    {
+    };
+    Train(std::string train_name) : train_name(train_name)
+    {
+    };
+    types::TrainId train_id;
+    std::string train_name;
+    std::vector<types::BlockId> block_occupancy;
+    types::BlockId current_position;
+    types::MetersPerSecond suggested_speed;
+    std::queue<types::BlockId> authority;
+    std::vector<DestinationAndArrivalTime> destination_list;
+    static types::TrainId last_id;
+
+    static types::TrainId GetNextId()
+    {
+        static types::TrainId last_id = 1;
+        return last_id++;
+    }
+};
 
 class Ctc
 {
     public:
-        void SetTrackLayout(const std::vector<types::Block> &blocks);
-        void SetSchedule(const types::TrainId train, const std::vector<DestinationAndArrivalTime> &schedule);
-        void SetManualMode(void);
+        /* Dispatch */
```

Comment:
Are all of these methods meant to be callable from outside the class?  Is there a need to call them outside the CTC class?  If not, make them private.